### ISSUE TITLE
Feature Request: AzureRM backend: Snapshot state files (blobs) before updating

### ISSUE THREAD

**Issue by kjhosein** [terraform-providers/terraform-provider-azurerm#1098]
### Terraform Version
Terraform v0.11.5
provider.azurerm v1.3.0

### Affected Resource(s)
- [azurerm backend](https://www.terraform.io/docs/backends/types/azurerm.html)

### Request Details
_Throwing this out there to start a conversation about it._ 
Over on the AWS side, currently if you enable S3 bucket versioning and store your state file remotely there, S3 will transparently take care of making a copy of the most recent version before Terraform uploads/changes the `terraform.state` file. 

From what I can tell in Azure blob storage, there is no exact equivalent (please let me know if I'm missing something). There are snapshots, but they need to be requested; they're not automatic. It would be great if TF had a toggle to request a snapshot be created prior to modifying the state file. Perhaps the `backend` might look like the code snippet below? 🤔 

#### Possible Terraform Configuration Files
```hcl
terraform {
  backend "azurerm" {
    storage_account_name = "sa1"
    container_name       = "azure-tf-state"
    key                  = "terraform.tfstate"
    versioning           = "true"
  }
}
```
Thanks!

**Comment by kjhosein** [comment]
BTW, for those interested in doing this external to Terraform, here is a _sample_ Azure CLI command:

`az storage blob snapshot --container-name azure-tf-state --name xyz/prod.terraform.tfstate `
[Reference: [Azure docs](https://docs.microsoft.com/en-us/cli/azure/storage/blob?view=azure-cli-latest#az-storage-blob-snapshot)]

You can throw this into a wrapper script just prior to your `tf apply`.

**Comment by szinck1** [comment]
+1 after nuking my tfstate and then realizing there's no backup ;)

**Comment by tombuildsstuff** [comment]
hey @kjhosein @szinck1 @bajorgensen 

Thanks for opening this issue :)

Since the Backend is housed within [the Terraform Core repository](https://github.com/hashicorp/terraform) - I've ported this issue over to [the main repository](https://github.com/hashicorp/terraform/issues/18284), where I've also added some additional context. As that's the better place to track this Feature Request - I'm going to close this issue in favour of that one.

Thanks!
